specs/capri/FileShare_Design.md

```diff
@@ -0,0 +1,741 @@
+* write a Manila driver and a default NFS/SMB driver
+* Provide CLI support for File Share service
+
+### Non-Goals
+
+* Provide support for multiple other file share protocols ex: GlusterFS, HDFS etc.. 
+* Supporting Replication and Data Protection for file share profile
+
+## Design Details
+
+1.  A user can request for action (let’s say: create a share) through
+     Northbound Plugins(NBP). We have already few plugins, but need to
+     modify for file share request. A User can also use CLI or dashboard.
+     So, all the actions(create, list, update, delete) should be supported
+     by CLI and dashboard.
+2. Now, NBP can communicate to OpenSDS API server.
+3. Based on user request File Share API will be selected.
+4. Internally, OpenSDS File Share API talk to controllers.
+5. Controllers will also maintain the File Share metadata in database.
+6. Controller communicates to Dock, specific driver to create a share.
+
+![FileShare Architecture Diagram](FileShare_Design.png?raw=true "File_Share_Design")
+
+#### Flow diagram:
+![FileShare Architecture Diagram](FileShare_Flow_diagram.png?raw=true "File_Share_Design")
+
+### REST API impact
+
+YES
+Fileshare API's needs to write and Existing Profile API's needs to be changed.
+
+### Data model impact
+
+YES
+
+### Security impact
+
+No
+
+### Other end user impact
+
+No
+
+## Use Cases
+
+Multiple users can access the same file share mount point.
+Data that would otherwise be duplicated on each client can be kept in a single location and accessed by client on the network.
+
+## Implementation
+This is the list of proposal for API
+
+##### 1. GET /v1beta/{tenantId}/file/shares
+* Lists all shares.
+ ##### Request
+ ```json
+"tenantId": "string"
+```
+  ##### Response parameters
+```json
+  “id” : “string”
+
+  “createdAt” : “2019-03-20T12:49:00.497Z”
+
+  “updatedAt” : “2019-03-20T12:49:00.497Z”
+
+  “tenantId” : “string”
+
+  “name” : “string”
+
+  “protocols”: "[]string"
+
+  “description” : “string”
+
+  “size” : “int64”
+
+  “availabilityZone”: “string”
+
+  “status”: “string”
+
+  “profileId”: “string”
+
+  “poolId”: “string”
+
+  “snapshotId”: “string”
+
+  “exportLocations” : "[]string"
+```
+##### 2. GET /v1beta/{tenantId}/file/shares/{shareId}
+* Shows details for a share.
+##### Request
+```json
+  “tenatId” : “string”
+
+  “shareId” : “string”
+
+  “exportLocations” : "[]string"
+```
+
+  ##### Response parameters
+  ```json
+  “id” : “string”
+
+  “protocols” : "[]string"
+
+  “createdAt” : “2019-03-20T12:49:00.497Z”
+
+  “updatedAt” : “2019-03-20T12:49:00.497Z”
+
+  “tenatId” : “string”
+
+  “userId” : “string“
+
+  “name” : “string”
+
+  “description” : “string”
+
+  “size” : “int64”
+
+  “availabilityZone”: “string”
+
+  “status”: “string”
+
+  “profileId”: “string”
+
+  “poolId”: “string”
+
+  “snapshotId”: “string”
+
+  “exportLocations” : “[]string”
+```
+
+##### 3.   POST /v1beta/{tenantId}/file/shares
+* Creates a share.
+##### Request
+```json
```

Comment:
Whether Add Access is supported when creating file share.

Reply:
No, Add access will only be supported once the file is created.